sql/mysqld.cc

```diff
@@ -5587,6 +5591,70 @@ static void test_lc_time_sz()
 }
 #endif//DBUG_OFF
 
+static void check_upgrade(void)
+{
+  File info_file;
+  char upgrade_from_version[1024]= {0};
+  size_t length;
+  ulong version_id;
+
+  uint major, minor, version;
+  char *point= upgrade_from_version, *end_point, *pos;
+
+  DBUG_ENTER("check_upgrade");
+
+  if (upgrade_info_file != 0)
+  {
+    fn_format(upgrade_info_file, "mysql_upgrade_info", mysql_real_data_home_ptr, "", MYF(0));
+    DBUG_PRINT("info", ("mysql_upgrade_info path: %s", mysql_real_data_home_ptr));
+  }
+
+  if ((info_file= my_open(upgrade_info_file, O_RDONLY, MYF(0))) < 0)
+  {
+    sql_print_information("Cannot open mysql_upgrade_info file: %s. Please run mariadb-upgrade.", upgrade_info_file);
+    DBUG_VOID_RETURN;
+  }
+
+  if (my_lock(info_file, F_RDLCK, 0, 1, MYF(0)))
+    goto end; // If Another process is holding a write lock avoid checking
+
+  (void) my_seek(info_file, 0, SEEK_SET, MYF(0));
+  /* We have -3 here to make conversion below safe (similar to mysql_upgrade) */
+  length= my_read(info_file, (uchar*) upgrade_from_version,
+                  sizeof(upgrade_from_version)-3, MYF(0));
+
+  /* Remove possible \ŋ that may end in output */
+  if ((pos= strchr(upgrade_from_version, '\n')))
+    *pos= 0;
+
+  if (!length)
+  {
+    sql_print_warning("mysql_upgrade_info file: %s is empty. Please run mariadb-upgrade.", upgrade_info_file);
+    goto end;
+  }
+
+  major=   (uint) strtoul(point, &end_point, 10);   point=end_point+1;
+  minor=   (uint) strtoul(point, &end_point, 10);   point=end_point+1;
+  version= (uint) strtoul(point, &end_point, 10);
+
+  if (major == minor && minor == version && version == 0)
+  {
+    sql_print_warning("The content of mysql_upgrade_info file: %s cannot be parsed. Please run mariadb-upgrade.", upgrade_info_file);
+    goto end;
+  }
+
+  version_id = (ulong) major * 10000L + (ulong)(minor * 100 + version);
+  if (MYSQL_VERSION_ID - version_id > 99)
```

Comment:
I'm not following the logic here. The expression is a major version behind right, how does that indicate that the upgrade began and didn't complete?

Reply:
Yes, as the content of the `mysql_upgrade_info` contains a major version older the one from the current binaries, it means that an upgrade have been started (i.e new binaries have been installed) but the `mysql_upgrade` tool haven't finished (or haven't ben run) yet. We can adjust the message if you think it might be misleading.

Reply:
Good explanation. Yes some rewording to represent accurate observations would be good.